### PR TITLE
Techdebt/noid/node14 npm7

### PR DESCRIPTION
**Pull request by marcoambrosini**
ref https://github.com/nextcloud/app-tutorial/pull/319#issuecomment-860745982

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -66,56 +66,26 @@
...
-		"eslint-plugin-jest": "^24.3.6",
-		"eslint-plugin-node": "^11.1.0",
-		"eslint-plugin-promise": "^4.3.1",
-		"eslint-plugin-standard": "^4.1.0",
-		"eslint-plugin-vue": "^5.2.3",
 		"jest-localstorage-mock": "^2.4.14",
 		"jest-mock-axios": "^4.4.0",
 		"jest-mock-console": "^1.1.0",
-		"node-sass": "^5.0.0",
 		"regenerator-runtime": "^0.13.7",
```
Comment (skjnldsv) [inline package.json:81]: should not be needed anymore

File: package.json [hunk 2]
```diff
@@ -66,56 +66,26 @@
...
+		"@nextcloud/webpack-vue-config": "^4.0.3",
 		"@vue/cli-plugin-unit-jest": "^4.5.13",
 		"@vue/cli-service": "^4.5.13",
 		"@vue/test-utils": "^1.2.0",
-		"babel-eslint": "^10.1.0",
-		"babel-loader": "^8.2.2",
 		"babel-loader-exclude-node-modules-except": "^1.2.1",
-		"babel-plugin-add-module-exports": "^1.0.4",
-		"css-loader": "^5.0.1",
-		"eslint": "^5.16.0",
```
Comment (PVince81) [inline package.json:84]: why are we getting rid of eslint ?

File: package.json [hunk 3]
```diff
@@ -66,56 +66,26 @@
...
-		"sass": "^1.34.1",
-		"sass-loader": "^10.1.1",
-		"style-loader": "^2.0.0",
-		"stylelint": "^13.13.1",
-		"stylelint-config-recommended-scss": "^4.2.0",
-		"stylelint-scss": "^3.19.0",
-		"stylelint-webpack-plugin": "^2.1.1",
-		"url-loader": "^4.1.1",
-		"vue-loader": "^15.9.7",
-		"vue-template-compiler": "^2.6.14",
```
Comment (PVince81) [inline package.json:108]: why remove this ? the jest test failure is expecting this lib to exist

Comment (PVince81) [reply package.json:108]: okay, I see, it becomes an implicit dep: https://github.com/nextcloud/webpack-vue-config/blob/master/package.json#L39

not sure why it's not found, probably Jest needs some extras as it doesn't use webpack

File: package.json (hunk 2 above)
Comment (skjnldsv) [reply package.json:84]: Some reading: https://github.com/nextcloud/app-tutorial/pull/319 ;)

File: .github/workflows/lint.yml [hunk 4]
```diff
@@ -55,18 +55,17 @@ jobs:
   node:
     runs-on: ubuntu-latest
 
-    strategy:
-      matrix:
-        node-versions: [12.x]
-
     name: eslint node${{ matrix.node-versions }}
```
Comment (nickvergessen) [inline .github/workflows/lint.yml:58]:
```suggestion
    name: eslint node
```

File: .github/workflows/lint.yml [hunk 5]
```diff
@@ -77,18 +76,17 @@ jobs:
   stylelint:
     runs-on: ubuntu-latest
 
-    strategy:
-      matrix:
-        node-versions: [12.x]
-
     name: stylelint node${{ matrix.node-versions }}
```
Comment (nickvergessen) [inline .github/workflows/lint.yml:79]:
```suggestion
    name: stylelint node
```

File: .github/workflows/lint.yml [hunk 6]
```diff
@@ -77,18 +76,17 @@ jobs:
...
     steps:
       - uses: actions/checkout@v2
 
       - name: Set up node ${{ matrix.node-versions }}
```
Comment (nickvergessen) [inline .github/workflows/lint.yml:83]:
```suggestion
      - name: Set up node
```

**Comment by PVince81** [comment]
@nickvergessen thanks for the suggestions. I already changed this in parallel, strange that Github doesn't say anything about this.

Now the errors are about actual code to adjust to satisfy eslint

**Comment by PVince81** [comment]
@marcoambrosini the jest stuff is fixed as you asked, do you want to take care of the reported eslint issues ?

**Comment by marcoambrosini** [comment]
> @marcoambrosini the jest stuff is fixed as you asked, do you want to take care of the reported eslint issues ?

Thanks @PVince81, yes I will!

**Comment by nickvergessen** [comment]
> do you want to take care of the reported eslint issues ?

We talked about it with skjnldsv yesterday and it might also be okay to add a temporary exception to the eslint as it's just a different prefered coding style

**Comment by skjnldsv** [comment]
> We talked about it with skjnldsv yesterday and it might also be okay to add a temporary exception to the eslint as it's just a different prefered coding style

Yes, requires investigation though, this is an error, not a warning :grin:

**Comment by marcoambrosini** [comment]
> Yes, requires investigation though, this is an error, not a warning 😁

Should be a super easy fix, on it

File: src/components/CallView/Grid/Grid.vue [hunk 7]
```diff
@@ -475,7 +476,7 @@ export default {
 
 	watch: {
 		// If the video array size changes, rebuild the grid
-		'videos.length': function() {
+		'videos.length'() {
```
Comment (skjnldsv) [inline src/components/CallView/Grid/Grid.vue:479]: Does this even works?
Can you watch length of arrays in vue watchers? 😮

Comment (marcoambrosini) [reply src/components/CallView/Grid/Grid.vue:479]: Seem to work fine, the grid rebuilds when someone joins :)

File: src/components/CallView/CallView.vue [hunk 8]
```diff
@@ -390,11 +391,15 @@ export default {
 		EventBus.$on('refreshPeerList', this.debounceFetchPeers)
 
 		callParticipantCollection.on('remove', this._lowerHandWhenParticipantLeaves)
+
+		subscribe('toggleVideo', this.handleToggleVideo)
```
Comment (skjnldsv) [inline src/components/CallView/CallView.vue:395]:
```suggestion
		subscribe('talk:video:toggled', this.handleToggleVideo)
```

Naming standard?

**Comment by marcoambrosini** [comment]
Ok it should™ all be good now, reviews plz!

File: package.json [hunk 9]
```diff
@@ -66,56 +66,26 @@
...
-		"webpack-cli": "^4.7.0",
-		"webpack-merge": "^5.7.3",
 		"webpack-node-externals": "^3.0.0"
 	},
 	"browserslist": [
 		"extends @nextcloud/browserslist-config"
 	],
 	"engines": {
-		"node": ">=10.0.0"
+		"node": ">=14.0.0"
```
Comment (nickvergessen) [inline package.json:88]: apparently we can also specify npm 7 here
https://github.com/nextcloud/photos/commit/c78052cc9f611c31e01ab98b834fc4101b0c5ad2

Comment (skjnldsv) [reply package.json]: https://docs.npmjs.com/cli/v6/configuring-npm/package-json#engines

File: package.json [hunk 10]
```diff
@@ -66,56 +66,27 @@
...
-		"webpack-merge": "^5.7.3",
 		"webpack-node-externals": "^3.0.0"
 	},
 	"browserslist": [
 		"extends @nextcloud/browserslist-config"
 	],
 	"engines": {
-		"node": ">=10.0.0"
+		"node": ">=14.0.0",
+		"npm": ">=7.0.0"
```
[the remark this answers is not in the document]
Comment (marcoambrosini) [reply package.json:89]: I'll rebase then

**Review by nickvergessen — CHANGES_REQUESTED** [review]
Seems to break calls completely

* can't see my own video
* sharing the screen doesn't work anymore

```
 [Vue warn]: Error in callback for watcher "localMediaModel.attributes.localStream": "TypeError: adapter.browserDetails is undefined"

found in

---> <LocalVideo> at src/components/CallView/shared/LocalVideo.vue
       <Grid> at src/components/CallView/Grid/Grid.vue
         <CallView> at src/components/CallView/CallView.vue
           <MainView> at src/views/MainView.vue
             <AppContent>
               <Content>
                 <App> at src/App.vue
                   <Root> vue.runtime.esm.js:619
```
```
TypeError: adapter.browserDetails is undefined
    exports attachmediastream.js:41
    _setLocalStream LocalVideo.vue:261
    localStream LocalVideo.vue:203
    VueJS 5
```

**Review by nickvergessen — APPROVED** [review]

File: src/components/CallView/CallView.vue [hunk 11]
```diff
@@ -620,6 +625,11 @@ export default {
 				console.error(exception)
 			}
 		}, 1500),
+
+		// Toggles videos on and off
+		handleToggleVideo({ peerId, value }) {
+			this.sharedDatas[peerId].videoEnabled = value
```
Comment (nickvergessen) [inline src/components/CallView/CallView.vue:631]: https://github.com/nextcloud/spreed/issues/5817

**Pull request merged**